### ISSUE TITLE
No approval email to booker

### ISSUE THREAD

**Issue by campbell-m** [meeting-room-booking-system/mrbs-code#3605]
(Transferred from the general mailing list)

I'm setting up MRBS and I’m having some difficult with email notifications. It seems that all works well when user makes a booking, the booker and admin gets an email. However when admin approves the booking the booker is not notified. It also happens when the admin edit/delete one previous approved booking. This are options what I think are relevant: 

~~~php
// WHO TO EMAIL
$mail_settings['admin_on_bookings']      = false;  // the addresses defined by $mail_settings['recipients'] below
$mail_settings['area_admin_on_bookings'] = false;  // the area administrator
$mail_settings['room_admin_on_bookings'] = true;  // the room administrator
$mail_settings['booker']                 = true;  // the person making the booking
$mail_settings['book_admin_on_approval'] = true;  // the booking administrator when booking approval is enabled

// WHEN TO EMAIL
$mail_settings['on_new']    = true;   // when an entry is created
$mail_settings['on_change'] = true;  // when an entry is changed
$mail_settings['on_delete'] = true;  // when an entry is deleted
~~~

Can anyone help figuring out what I’m doing wrong?

Thank you,
Luís Martins

**Comment by campbell-m** [comment]
I've just tested it and it works OK for me.  If you haven't already then try setting

~~~php
$mail_settings['use_from_for_all_mail'] = true;
~~~

**Comment by lordgabriel** [comment]
I start the configuration from the beginning, step by step, and I think what is causing the problem. With the db auth config all emails notification were ok but after I use Shibboleth Authentication handler for MRBS. from Ondřej Votava and Martin Samek it stopped working. I will try to figure out why...

Turning on the email debug it shows that the 'To' field have 'undisclosed-recipients:;' instead of the booker address

 I will try to figure out why...

**Comment by lordgabriel** [comment]
I cannot figure out why the approval email is not beeing sent when I use Shibbolet Authentication scheme.

I'm using this [Shibboleth Auth scheme](https://gitlab.fel.cvut.cz/fel-svti-aplikace/mrbs-shibboleth/-/tree/master)

Appreciate if anyone can help. Thank you

**Comment by campbell-m** [comment]
I think it's probably because the Shibboleth authentication scheme doesn't have a way of retrieving the details of a user who isn't the currently logged in user.  I assume that you have the MRBS admin email address in your MRBS config file, so when the user makes a booking it can retrieve both the email address of the booker (current user) and also the admin (from the config file).   However when the admin approves it the system can't get the booker's details.

If you know what the email address of all users is given their username (eg it's '<username>@domain.com') then you could modify the `getEmail()` method in Auth\AuthShibboleth.php.  Otherwise you could ask the authors for help.

**Comment by lordgabriel** [comment]
Makes sense.  I assumed (wrong) that the email of the user was recorded with the booking info. Thank you

**Comment by campbell-m** [comment]
Yes, the booking info just has the username.  That's because other details, eg the display name and email address, may change after the booking is made.

**Comment by codingmona** [comment]
Hello,
Was anyone able to figure out a solution for this? We are using simplesamlphp to login to MRBS so probably only username is stored in the booking details.

**Comment by mbrucker-kiel** [comment]
Anyone found a solution yet?

**Comment by mbrucker-kiel** [comment]
Hi @campbell-m  👋

I’m still hitting this issue in #3605 and would really like to get it fixed properly.
If you’re open to it, I can put a **paid bounty** on it.

I can offer **150–400 €** for a clean fix that gets merged.

What I need:
- issue #3605 fixed reliably (mail notifications behaving correctly)
- no regressions in related notification behavior
- ideally tests/docs updated if needed
- important for me: a **working integration for SAML (via Keycloak)** in my setup

One important point: I need a invoice for internal accounting, otherwise I can’t process payment.

If this is okay, I’m happy to coordinate details (scope, timeline, payment method) and follow whatever process you prefer.
Thanks a lot 🙏

**Comment by campbell-m** [comment]
It's either a simple fix and I'd be happy to help you out for nothing, or else it's going to be much more complicated and I wouldn't have time to look at it.  What have you got `$auth['type']` and `$auth['session']` set to?

**Comment by mbrucker-kiel** [comment]
Booth is set to:
```
$auth['type'] = 'saml';
$auth['session'] = 'saml';
```

Mapping the User-Email via:
```
$auth['saml']['attr']['username'] = 'username';
$auth['saml']['attr']['mail'] = 'email';
$auth['saml']['attr']['givenName'] = 'firstName';
$auth['saml']['attr']['surname'] = 'lastName';
$auth['saml']['admin']['groups'] = ['mrbs-admin'];
```

**Comment by campbell-m** [comment]
Sorry, I think you're going to have to write/find someone to write the code.

The problem is that the 'saml' authentication as it stands can just get the details (email address, display name, etc.) of the currently logged in user.  So when the admin approves a booking, it cannot get the booker's details so can't send them an email.  Looking at the Keycloak website there is an API for getting a user's details, so it should be possible, but is probably specifc to Keycloak (though I haven't investigated it in depth).

What's needed is a new 'keycloak' authentication type and the code would look something like this:

~~~php
<?php
namespace MRBS\Auth;

use MRBS\User;

class AuthKeycloak extends AuthSaml
{
  protected function getUserFresh(string $username) : ?User
  {
    // Use the Keycloak API to get the details for $username and return a User object
  }
}
~~~

In the longer term, it looks as though Keycloak also offers an OAuth2 API (maybe that's the one I was looking at).   MRBS really needs an OAuth2 authentication capability, which would be quite a bit of work, but if it had this then extending it to Keycloak should be relatively simple.

**Comment by campbell-m** [comment]
> In the longer term, it looks as though Keycloak also offers an OAuth2 API (maybe that's the one I was looking at). MRBS really needs an OAuth2 authentication capability, which would be quite a bit of work, but if it had this then extending it to Keycloak should be relatively simple.

See #4000.

**Comment by 3157696** [comment]
Everyone 
When booking a meeting, does the booker also invite the list of participants? Have the system send them an email, as there is no relevant input box on the UI
Similar to using the meeting invitation function in Outlook

**Comment by campbell-m** [comment]
> When booking a meeting, does the booker also invite the list of participants?

No, it doesn't.
